File: backend/app/schemas/admin_master.py
"""Admin Master Data management schemas."""
from datetime import datetime
from pydantic import BaseModel, Field, ConfigDict


# ---------------------------------------------------------------------------
# Lines
# ---------------------------------------------------------------------------

class LineCreate(BaseModel):
    line_code: str = Field(..., max_length=50)
    line_name: str = Field(..., max_length=100)


class LineUpdate(BaseModel):
    line_code: str | None = Field(None, max_length=50)
    line_name: str | None = Field(None, max_length=100)


class LineResponse(BaseModel):
    id: int
    line_code: str
    line_name: str
    created_at: datetime
    product_count: int = 0

    model_config = ConfigDict(from_attributes=True)


# ---------------------------------------------------------------------------
# Products
# ---------------------------------------------------------------------------

class ProductCreate(BaseModel):
    product_name: str = Field(..., max_length=100)
    description: str | None = None
    line_id: int | None = None
    part_id: str | None = Field(None, max_length=100)


class ProductUpdate(BaseModel):
    product_name: str | None = Field(None, max_length=100)
    description: str | None = None
    line_id: int | None = None
    part_id: str | None = None


class ProductResponse(BaseModel):
    id: int
    product_name: str
    description: str | None = None
    line_id: int | None = None
    line_name: str | None = None
    part_id: str | None = None
    created_at: datetime

    model_config = ConfigDict(from_attributes=True)


# ---------------------------------------------------------------------------
# Layers
# ---------------------------------------------------------------------------

class LayerCreate(BaseModel):
    layer_name: str = Field(..., max_length=100)
    step_seq: str = Field(..., max_length=10)
    layer_number: str = Field(..., max_length=10)
    sort_order: int = 0


class LayerUpdate(BaseModel):
    layer_name: str | None = Field(None, max_length=100)
    step_seq: str | None = Field(None, max_length=10)
    layer_number: str | None = Field(None, max_length=10)
    sort_order: int | None = None


class LayerResponse(BaseModel):
    id: int
    layer_name: str
    step_seq: str
    layer_number: str
    sort_order: int
    created_at: datetime

    model_config = ConfigDict(from_attributes=True)


class LayerReorderRequest(BaseModel):
    ordered_ids: list[int]


# ---------------------------------------------------------------------------
# Column metadata
# ---------------------------------------------------------------------------

class ColumnMetadataUpdate(BaseModel):
    display_name: str | None = Field(None, max_length=200)
    unit: str | None = None
    is_required: bool | None = None


class ColumnMetadataResponse(BaseModel):
    id: int
    column_name: str
    display_name: str
    category_code: str | None = None
    data_type: str
    unit: str | None = None
    is_required: bool
    sort_order: int

    model_config = ConfigDict(from_attributes=True)


# ---------------------------------------------------------------------------
# Column create
# ---------------------------------------------------------------------------

class ColumnCreateRequest(BaseModel):
    column_name: str = Field(..., max_length=100)
    display_name: str = Field(..., max_length=200)
    category_id: int
    data_type: str = Field(..., max_length=20)  # integer / float / string / select
    unit: str | None = None
    is_required: bool = False
    select_options: list[str] | None = None


# ---------------------------------------------------------------------------
# Category
# ---------------------------------------------------------------------------

class CategoryCreateRequest(BaseModel):
    category_code: str = Field(..., max_length=10)
    category_name: str = Field(..., max_length=50)


class CategoryUpdate(BaseModel):
    category_name: str | None = Field(None, max_length=50)


class CategoryResponse(BaseModel):
    id: int
    category_code: str
    category_name: str
    sort_order: int
    column_count: int = 0

    model_config = ConfigDict(from_attributes=True)


class CategoryReorderRequest(BaseModel):
    ordered_ids: list[int]


# ---------------------------------------------------------------------------
# Equipments
# ---------------------------------------------------------------------------

class EquipmentCreate(BaseModel):
    line_id: int
    equipment_name: str = Field(..., max_length=100)
    equipment_model: str | None = Field(None, max_length=100)
    prc: str | None = Field(None, max_length=50)
    ip: str | None = Field(None, max_length=45)
    ftp_id: str | None = Field(None, max_length=100)
    ftp_pw: str | None = Field(None, max_length=255)
    is_active: bool = True
    sort_order: int = 0


class EquipmentUpdate(BaseModel):
    equipment_name: str | None = Field(None, max_length=100)
    equipment_model: str | None = None
    prc: str | None = None
    ip: str | None = None
    ftp_id: str | None = None
    ftp_pw: str | None = None  # write-only, not returned in response
    is_active: bool | None = None
    sort_order: int | None = None


class EquipmentResponse(BaseModel):
    id: int
    line_id: int
    line_name: str = ""
    equipment_name: str
    equipment_model: str | None = None
    prc: str | None = None
    ip: str | None = None
    ftp_id: str | None = None
    # ftp_pw EXCLUDED - write-only field (REQ-EQP-017)
    is_active: bool
    sort_order: int

    model_config = ConfigDict(from_attributes=True)


class EquipmentPublicResponse(BaseModel):
    id: int
    equipment_name: str
    equipment_model: str | None = None

    model_config = ConfigDict(from_attributes=True)


class EquipmentReorderRequest(BaseModel):
    ordered_ids: list[int]

File: frontend/src/api/adminMaster.ts
import client from './client'

// ========== Types ==========

export interface LineResponse {
  id: number
  line_code: string
  line_name: string
  created_at: string
  product_count: number
}

export interface LineCreate {
  line_code: string
  line_name: string
}

export interface ProductResponse {
  id: number
  product_name: string
  description: string | null
  line_id: number | null
  line_name: string | null
  part_id: string | null
  created_at: string
}

export interface ProductCreate {
  product_name: string
  description?: string | null
  line_id?: number | null
  part_id?: string | null
}

export interface LayerResponse {
  id: number
  layer_name: string
  step_seq: string
  layer_number: number
  sort_order: number
  created_at: string
}

export interface LayerCreate {
  layer_name: string
  step_seq: string
  layer_number: number
  sort_order: number
}

export interface ColumnMetadataResponse {
  id: number
  column_name: string
  display_name: string
  category_id: number | null
  category_code: string | null
  data_type: string
  unit: string | null
  is_required: boolean
  use_yn: boolean
  sort_order: number
}

export interface ColumnMetadataUpdate {
  display_name?: string
  unit?: string | null
  is_required?: boolean
  use_yn?: boolean
}

export interface ColumnCreateRequest {
  column_name: string
  display_name: string
  category_id: number
  data_type: string
  unit?: string | null
  is_required?: boolean
  use_yn?: boolean
  select_options?: string[] | null
}

export interface CategoryResponse {
  id: number
  category_code: string
  category_name: string
  sort_order: number
  column_count: number
}

export interface CategoryCreateRequest {
  category_code: string
  category_name: string
}

export interface CategoryUpdate {
  category_name?: string
  sort_order?: number
}

// ========== Lines ==========

export async function fetchAdminLines(): Promise<LineResponse[]> {
  const { data } = await client.get<LineResponse[]>('/admin/lines')
  return data
}

export async function createLine(payload: LineCreate): Promise<LineResponse> {
  const { data } = await client.post<LineResponse>('/admin/lines', payload)
  return data
}

export async function updateLine(id: number, payload: Partial<LineCreate>): Promise<LineResponse> {
  const { data } = await client.put<LineResponse>(`/admin/lines/${id}`, payload)
  return data
}

export async function deleteLine(id: number): Promise<void> {
  await client.delete(`/admin/lines/${id}`)
}

// ========== Products ==========

export async function fetchAdminProducts(lineId?: number): Promise<ProductResponse[]> {
  const { data } = await client.get<ProductResponse[]>('/admin/products', {
    params: lineId ? { line_id: lineId } : undefined,
  })
  return data
}

export async function createProduct(payload: ProductCreate): Promise<ProductResponse> {
  const { data } = await client.post<ProductResponse>('/admin/products', payload)
  return data
}

export async function updateProduct(
  id: number,
  payload: Partial<ProductCreate>
): Promise<ProductResponse> {
  const { data } = await client.put<ProductResponse>(`/admin/products/${id}`, payload)
  return data
}

export async function deleteProduct(id: number): Promise<void> {
  await client.delete(`/admin/products/${id}`)
}

// ========== Layers ==========

export async function fetchAdminLayers(): Promise<LayerResponse[]> {
  const { data } = await client.get<LayerResponse[]>('/admin/layers')
  return data
}

export async function createLayer(payload: LayerCreate): Promise<LayerResponse> {
  const { data } = await client.post<LayerResponse>('/admin/layers', payload)
  return data
}

export async function updateLayer(
  id: number,
  payload: Partial<LayerCreate>
): Promise<LayerResponse> {
  const { data } = await client.put<LayerResponse>(`/admin/layers/${id}`, payload)
  return data
}

export async function deleteLayer(id: number): Promise<void> {
  await client.delete(`/admin/layers/${id}`)
}

export async function reorderLayers(orderedIds: number[]): Promise<LayerResponse[]> {
  const { data } = await client.put<LayerResponse[]>('/admin/layers/reorder', {
    ordered_ids: orderedIds,
  })
  return data
}

// ========== Columns ==========

export async function fetchAdminColumns(): Promise<ColumnMetadataResponse[]> {
  // 관리자용: use_yn 무관 전체 컬럼 조회 (비공개 컬럼 관리 화면용)
  const { data } = await client.get<ColumnMetadataResponse[]>('/admin/columns')
  return data
}

export async function createColumn(payload: ColumnCreateRequest): Promise<ColumnMetadataResponse> {
  const { data } = await client.post<ColumnMetadataResponse>('/admin/columns', payload)
  return data
}

export async function updateColumnMetadata(
  id: number,
  payload: ColumnMetadataUpdate
): Promise<ColumnMetadataResponse> {
  const { data } = await client.put<ColumnMetadataResponse>(
    `/admin/columns/${id}/metadata`,
    payload
  )
  return data
}

export async function deleteColumn(id: number): Promise<void> {
  await client.delete(`/admin/columns/${id}`)
}

// ========== Categories ==========

export async function fetchCategories(): Promise<CategoryResponse[]> {
  const { data } = await client.get<CategoryResponse[]>('/admin/categories')
  return data
}

export async function createCategory(payload: CategoryCreateRequest): Promise<CategoryResponse> {
  const { data } = await client.post<CategoryResponse>('/admin/categories', payload)
  return data
}

export async function updateCategory(
  id: number,
  payload: CategoryUpdate
): Promise<CategoryResponse> {
  const { data } = await client.put<CategoryResponse>(`/admin/categories/${id}`, payload)
  return data
}

export async function deleteCategory(id: number): Promise<void> {
  await client.delete(`/admin/categories/${id}`)
}

export async function reorderCategories(orderedIds: number[]): Promise<CategoryResponse[]> {
  const { data } = await client.put<CategoryResponse[]>('/admin/categories/reorder', {
    ordered_ids: orderedIds,
  })
  return data
}

// ========== Equipments ==========

export interface EquipmentResponse {
  id: number
  line_id: number
  line_name: string
  equipment_name: string
  equipment_model: string | null
  prc: string | null
  ip: string | null
  ftp_id: string | null
  is_active: boolean
  sort_order: number
}

export interface EquipmentCreate {
  line_id: number
  equipment_name: string
  equipment_model?: string
  prc?: string
  ip?: string
  ftp_id?: string
  ftp_pw?: string
  is_active?: boolean
  sort_order?: number
}

export async function fetchAdminEquipments(lineId?: number): Promise<EquipmentResponse[]> {
  const { data } = await client.get<EquipmentResponse[]>('/admin/equipments', {
    params: lineId ? { line_id: lineId } : undefined,
  })
  return data
}

export async function createEquipment(payload: EquipmentCreate): Promise<EquipmentResponse> {
  const { data } = await client.post<EquipmentResponse>('/admin/equipments', payload)
  return data
}

export async function updateEquipment(
  id: number,
  payload: Partial<EquipmentCreate>
): Promise<EquipmentResponse> {
  const { data } = await client.put<EquipmentResponse>(`/admin/equipments/${id}`, payload)
  return data
}

export async function deleteEquipment(id: number): Promise<void> {
  await client.delete(`/admin/equipments/${id}`)
}
